aten/src/ATen/DeviceGuard.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <ATen/Device.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+
+namespace at {
+/// RAII guard that sets a certain default GPU index in its constructor, and
+/// changes it back to the device that was originally active upon destruction.
+///
+/// The index is always reset to the one that was active at the time of
+/// construction of the guard. Even if you `set_index` after construction, the
+/// destructor will still reset the index to the one that was active at
+/// construction time.
+///
+/// Legacy constructors and setters are kept around that accept -1 as the device
+/// index, which is equivalent to `at::nullopt`. All new code should prefer the
+/// latter.
+struct DeviceGuard {
+  /// Default constructor, does nothing.
+  DeviceGuard();
+
+  DeviceGuard(at::optional<int32_t> index);
+
+  /// Legacy constructor that accepts -1 as the device index and turns it into
+  /// `at::nullopt`.
+  /* deprecated */ DeviceGuard(int32_t index);
+
+  /// Sets the device to the index on which the given tensor is located.
+  explicit DeviceGuard(const Tensor& tensor);
+
+  /// Sets the device to the index on which the first tensor in the list is
+  /// located. If the list is empty, does nothing.
+  explicit DeviceGuard(const TensorList& tensors);
```

Comment:
can we just do this explicitly?  IMO it's better to make it obvious what's happening at the call site than not.

Reply:
I think I looked into this originally, but the issue is this constructor is used in generated code in `python_torch_functions_dispatch.h`, and it's much more convenient to just use the same code for all dispatch functions instead of changing the codegen